src/mbgl/text/glyph_set.cpp

```diff
@@ -88,82 +85,100 @@ void justifyLine(std::vector<PositionedGlyph> &positionedGlyphs, const std::map<
-                line++;
-            }
-
-            // Ideographic characters, spaces, and word-breaking punctuation that often appear without surrounding spaces.
-            if (useBalancedIdeographicBreaking
-                || util::i18n::allowsWordBreaking(shape.glyph)
-                || util::i18n::allowsIdeographicBreaking(shape.glyph)) {
-                lastSafeBreak = i;
-            }
+        // Ideographic characters, spaces, and word-breaking punctuation that often appear without surrounding spaces.
+        if (useBalancedIdeographicBreaking
+            || util::i18n::allowsWordBreaking(glyph.id)
+            || util::i18n::allowsIdeographicBreaking(glyph.id)) {
+            lastSafeBreak = i;
+            lastSafeBreakX = currentX;
         }
     }
 
-    const PositionedGlyph& lastPositionedGlyph = positionedGlyphs.back();
-    const auto lastGlyphIt = sdfs.find(lastPositionedGlyph.glyph);
-    assert(lastGlyphIt != sdfs.end());
-    const uint32_t lastLineLength = lastPositionedGlyph.x + lastGlyphIt->second.metrics.advance;
-    maxLineLength = std::max(maxLineLength, lastLineLength);
-
-    const uint32_t height = (line + 1) * lineHeight;
+    return lineBreakPoints;
+}
 
-    justifyLine(positionedGlyphs, sdfs, lineStartIndex, uint32_t(positionedGlyphs.size()) - 1, justify);
-    align(shaping, justify, horizontalAlign, verticalAlign, maxLineLength, lineHeight, line, translate);
+void GlyphSet::shapeLines(Shaping& shaping, const std::vector<std::u16string>& lines,
+                             const float spacing, const float lineHeight, const float horizontalAlign, const float verticalAlign,
+                             const float justify, const Point<float> &translate) const {
 
+    // the y offset *should* be part of the font metadata
+    const int32_t yOffset = -17;
+    
+    float x = 0;
+    float y = yOffset;
+    
+    float maxLineLength = 0;
+    
+    for (std::u16string line : lines)
+    {
+        // Filter out any missing glyphs ahead of time, after this assume we have all glyphs
+        line.erase(std::remove_if(line.begin(),
+                                  line.end(),
+                                  [ & ] (const char16_t& chr) -> bool { auto it = sdfs.find(chr); return it == sdfs.end(); }),
+                   line.end());
+        
+        if (line.empty())
+            continue;
+        
+        uint32_t lineStartIndex = shaping.positionedGlyphs.size();
+        for (char16_t chr : line)
+        {
+            const SDFGlyph& glyph = sdfs.find(chr)->second;
+            shaping.positionedGlyphs.emplace_back(chr, x, y);
+            x += glyph.metrics.advance + spacing;
+        }
+        
+        // Collapse invisible characters.
+        uint32_t lineEndIndex = shaping.positionedGlyphs.size() - 1;
+        // TODO refactor: Seems like isVisible should be called isInvisible? Also there's an edge case here with one character invisible lines
```

Comment:
I had to torture the map a little to get it to happen, but there is a crash bug here: if you have labels that include trailing line breaks (you can copy and paste them into Studio even though you can't enter them directly), you can end up with "lines" that just have a single line break on them. In that case lineEnd-- wraps around to 2^32.